### PR TITLE
Add rustX check to codeblock attributes lint

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
We discovered this issue [here](https://github.com/rust-lang/rust/pull/118802#discussion_r1421815943).

I assume that the issue will be present in other places outside of the compiler so it's worth adding a check for it.

First commit is just a small cleanup about variables creation which was a bit strange (at least more than necessary).

r? @notriddle

### CODE REVIEW

File: src/librustdoc/html/markdown.rs [hunk 1]
```diff
@@ -1293,6 +1284,16 @@ impl LangString {
                     LangStringToken::LangToken(x) if x.starts_with("edition") => {
                         data.edition = x[7..].parse::<Edition>().ok();
                     }
+                    LangStringToken::LangToken(x) if x.starts_with("rust") => {
+                        if x[4..].parse::<Edition>().is_ok()
+                            && let Some(extra) = extra
+                        {
+                            extra.error_invalid_codeblock_attr(format!(
+                                "unknown attribute `{x}`. Did you mean `edition{}`?",
```
Comment (Urgau) [inline src/librustdoc/html/markdown.rs:1292]: The rustc-dev-guide says ["did you mean" should be avoided](https://rustc-dev-guide.rust-lang.org/diagnostics.html#suggestion-style-guide):
> Suggestions should not be a question. In particular, language like "did you mean" should be avoided. Sometimes, it's unclear why a particular suggestion is being made. In these cases, it's better to be upfront about what the suggestion is.
>
> Compare "did you mean: Foo" vs. "there is a struct with a similar name: Foo".

Comment (GuillaumeGomez) [reply src/librustdoc/html/markdown.rs:1292]: Oh nice! I'll update all the suggestions then. :)

**Comment by GuillaumeGomez** [comment]
Thanks for the suggestion @Urgau!

File: tests/rustdoc-ui/doctest/check-attr-test.stderr [hunk 2]
```diff
@@ -1,4 +1,4 @@
-error: unknown attribute `compile-fail`. Did you mean `compile_fail`?
+error: unknown attribute `compile-fail`. There is an attribute with a similar name: `compile_fail`?
```
Comment (Urgau) [inline tests/rustdoc-ui/doctest/check-attr-test.stderr:1]: Why is the suggestion in the error message? I think it should be in a help suggestion (maybe even a machine-application suggestion).

Comment (Urgau) [reply tests/rustdoc-ui/doctest/check-attr-test.stderr:1]: Also, it's not a question anymore.
```suggestion
error: unknown attribute `compile-fail`. There is an attribute with a similar name: `compile_fail`
```

Comment (GuillaumeGomez) [reply tests/rustdoc-ui/doctest/check-attr-test.stderr:1]: Good point, thanks!

Comment (GuillaumeGomez) [reply tests/rustdoc-ui/doctest/check-attr-test.stderr]: And done! :)

**Comment by GuillaumeGomez** [comment]
Fixed the error.

**Review by notriddle — APPROVED** [review]
Other than the above comment about combining multiple mistakes, this seems like a great diagnostic improvement. Thanks!

Add a test case for combining them, then r=me

File: tests/rustdoc-ui/lints/check-attr.stderr [hunk 3]
```diff
@@ -169,7 +180,21 @@ LL | | /// boo
...
+LL | |
+LL | | ///
+LL | | /// ```rust2018
+LL | | /// boo
+LL | | /// ```
+   | |_______^
+   |
+   = help: there is an attribute with a similar name: `edition2018`
+
+error: aborting due to 13 previous errors
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply tests/rustdoc-ui/lints/check-attr.stderr:200]: This:

```
warning: unknown attribute `rust2018`
 --> foo.rs:3:1
  |
3 | / /**
4 | | ```rust2018 shouldpanic
5 | | panic!();
6 | | ```
7 | | */
  | |__^
  |
  = help: there is an attribute with a similar name: `edition2018`
  = note: `#[warn(rustdoc::invalid_codeblock_attributes)]` on by default

warning: unknown attribute `shouldpanic`
 --> foo.rs:3:1
  |
3 | / /**
4 | | ```rust2018 shouldpanic
5 | | panic!();
6 | | ```
7 | | */
  | |__^
  |
  = help: there is an attribute with a similar name: `should_panic`
  = help: the code block will either not be tested if not marked as a rust one or won't fail if it doesn't panic when running

warning: 2 warnings emitted
```

**Pull request merged**